experiments/yarn/src/main/scala/org/apache/gearpump/experiments/yarn/master/YarnApplicationMaster.scala

```diff
@@ -0,0 +1,324 @@
+    nodeManagerClient.startContainerAsync(container, containerContext)
+  }
+
+  private def requestWorkerContainers(): Unit = {
+    (1 to appConfig.getEnv(WORKER_CONTAINERS).toInt).foreach(requestId => {
+      resourceManagerClient ! ContainerRequestMessage(appConfig.getEnv(WORKER_MEMORY).toInt, appConfig.getEnv(WORKER_VCORES).toInt)
+    })
+  }
+
+  private def requestMasterContainers(registrationResponse: RegisterApplicationMasterResponse) = {
+    //TODO remove that or do something more with that
+    val previousContainersCount = registrationResponse.getContainersFromPreviousAttempts.size
+
+    LOG.info(s"Previous container count : $previousContainersCount")
+    if (previousContainersCount > 0) {
+      LOG.warn("Previous container count > 0, can't do anything with it")
+    }
+
+    LOG.info(s"GEARPUMPMASTER_CONTAINERS: ${appConfig.getEnv(GEARPUMPMASTER_CONTAINERS).toInt}")
+    (1 to appConfig.getEnv(GEARPUMPMASTER_CONTAINERS).toInt).foreach(requestId => {
+      resourceManagerClient ! ContainerRequestMessage(appConfig.getEnv(GEARPUMPMASTER_MEMORY).toInt, appConfig.getEnv(GEARPUMPMASTER_VCORES).toInt)
+    })
+
+  }
+
+  private def cleanUp(state: RMTerminalState): Boolean = {
+    LOG.info("Application completed. Stopping running containers")
+    nodeManagerClient.stop()
+    var success = true
+
+    val stats = state.containerStats
+    state match {
+      case error: RMError =>
+        LOG.info("Failed", error.throwable.getMessage)
+        val message = s"Failed. total=${appConfig.getEnv(WORKER_CONTAINERS).toInt}, completed=${stats.completed}, allocated=${stats.allocated}, failed=${stats.failed}"
+        resourceManagerClient ! AMStatusMessage(FinalApplicationStatus.FAILED, message, null)
+        success = false
+      case shutdownRequest: RMShutdownRequest =>
+        if (stats.failed == 0 && stats.completed == appConfig.getEnv(WORKER_CONTAINERS).toInt) {
+          val message = s"ShutdownRequest. total=${appConfig.getEnv(WORKER_CONTAINERS).toInt}, completed=${stats.completed}, allocated=${stats.allocated}, failed=${stats.failed}"
+          resourceManagerClient ! AMStatusMessage(FinalApplicationStatus.KILLED, message, null)
+          success = false
+        } else {
+          val message = s"ShutdownRequest. total=${appConfig.getEnv(WORKER_CONTAINERS).toInt}, completed=${stats.completed}, allocated=${stats.allocated}, failed=${stats.failed}"
+          resourceManagerClient ! AMStatusMessage(FinalApplicationStatus.FAILED, message, null)
+          success = false
+        }
+      case allRequestedContainersCompleted: RMAllRequestedContainersCompleted =>
+        val message = s"Diagnostics. total=${appConfig.getEnv(WORKER_CONTAINERS).toInt}, completed=${stats.completed}, allocated=${stats.allocated}, failed=${stats.failed}"
+        resourceManagerClient ! AMStatusMessage(FinalApplicationStatus.SUCCEEDED, message, null)
+        success = true
+    }
+
+    resourceManagerClient ! PoisonPill
+    success
```

Comment:
What the return value "success" here mean?
If "success" == false, it means cleanup failed?